### PR TITLE
feat(angular-ionic): add one signal directives

### PR DESCRIPTION
**Pull request by eesteban**
Add directives for:
- Initialize one-signal connection
- Connect/Disconnect users in one-signal

### CODE REVIEW

File: @santech/angular-ionic/testing/ionic/index.ts [hunk 1]
```diff
@@ -152,6 +165,27 @@ if (spyPlatform) {
   });
 }
 
+if (spyOneSignal) {
+  Object.defineProperty(spyOneSignal, 'OSInFocusDisplayOption', {
```
Comment (paulsouche) [inline @santech/angular-ionic/testing/ionic/index.ts]: Are these readonly ?

Comment (eesteban) [reply @santech/angular-ionic/testing/ionic/index.ts]: yes

File: @santech/angular-ionic/src/directives/one-signal.directive.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+  public subscriptionChanged = new EventEmitter<OSSubscriptionState>();
+
+  private _oneSignalService: OneSignalService;
+
+  constructor(oneSignalService: OneSignalService) {
+    this._oneSignalService = oneSignalService;
+  }
+
+  public ngOnInit() {
+    this._oneSignalService.init()
```
Comment (paulsouche) [inline @santech/angular-ionic/src/directives/one-signal.directive.ts]: return

File: @santech/angular-ionic/src/directives/one-signal-authenticated.directive.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+  private _oneSignalService: OneSignalService;
+
+  constructor(oneSignalService: OneSignalService) {
+    this._oneSignalService = oneSignalService;
+  }
+
+  public ngOnInit() {
+    return this._oneSignalService.enableNotifications()
+      .then((ids) => {
```
Comment (paulsouche) [inline @santech/angular-ionic/src/directives/one-signal-authenticated.directive.ts]: Emit ids anyway IOneSignalIds | undefined

File: @santech/angular-ionic/src/directives/one-signal-authenticated.directive.ts [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+        if (ids) {
+          this.onNotificationsEnabled.emit(ids);
+        }
+      });
```
Comment (paulsouche) [inline @santech/angular-ionic/src/directives/one-signal-authenticated.directive.ts]: What if this fail ? noop ?

File: @santech/angular-ionic/src/directives/one-signal.directive.spec.ts [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+
+  beforeEach(() => {
+    jest.resetAllMocks();
+
+    spyOneSignal.handleNotificationReceived.mockReturnValue(handleNotificationReceived);
+    spyOneSignal.handleNotificationOpened.mockReturnValue(handleNotificationOpened);
+    spyOneSignal.addSubscriptionObserver.mockReturnValue(addSubscriptionObserver);
+
+    TestBed
+    .configureTestingModule({
```
Comment (paulsouche) [inline @santech/angular-ionic/src/directives/one-signal.directive.spec.ts]: indentation

File: @santech/angular-ionic/src/directives/one-signal.directive.ts [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+          notificationReceived,
+          subscriptionChanged,
+        } = handlers;
+
+        this._subscriptions = [
+          notificationOpened.subscribe((value) => this.notificationOpened.emit(value)),
+          notificationReceived.subscribe((value) => this.notificationReceived.emit(value)),
+          subscriptionChanged.subscribe((value) => this.subscriptionChanged.emit(value)),
+        ];
+      });
```
Comment (paulsouche) [inline @santech/angular-ionic/src/directives/one-signal.directive.ts]: What if this fail ?

File: @santech/angular-ionic/src/directives/one-signal.directive.ts [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+          return;
+        }
+
+        const {
+          notificationOpened,
+          notificationReceived,
+          subscriptionChanged,
+        } = handlers;
+
+        this._subscriptions = [
```
Comment (paulsouche) [inline @santech/angular-ionic/src/directives/one-signal.directive.ts]: push ?

File: @santech/angular-ionic/src/services/one-signal.service.ts [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+    };
+  }
+
+  public async enableNotifications(): Promise<IOneSignalIds | undefined> {
+    const pt = await this._platform.ready();
+    if (pt !== cordovaPlatform) {
+      return;
+    }
+
+    const oneSignal =  this._oneSignal;
```
Comment (paulsouche) [inline @santech/angular-ionic/src/services/one-signal.service.ts]: indentation

File: @santech/angular-ionic/src/services/one-signal.service.ts [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+    oneSignal.endInit();
+    return {
+      notificationOpened: oneSignal.handleNotificationOpened(),
+      notificationReceived: oneSignal.handleNotificationReceived(),
+      subscriptionChanged: oneSignal.addSubscriptionObserver(),
+    };
+  }
+
+  public async enableNotifications(): Promise<IOneSignalIds | undefined> {
+    const pt = await this._platform.ready();
```
Comment (paulsouche) [inline @santech/angular-ionic/src/services/one-signal.service.ts]: You should extract a method

File: @santech/angular-ionic/src/services/one-signal.service.ts [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+    oneSignal.endInit();
+
+    return {
+      notificationOpened: oneSignal.handleNotificationOpened(),
+      notificationReceived: oneSignal.handleNotificationReceived(),
+      subscriptionChanged: oneSignal.addSubscriptionObserver(),
+    };
+  }
+
+  public enableNotifications(): Promise<IOneSignalIds> {
```
Comment (paulsouche) [inline @santech/angular-ionic/src/services/one-signal.service.ts]: async

File: @santech/angular-ionic/src/directives/one-signal.directive.ts [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    const {
+      notificationOpened,
+      notificationReceived,
+      subscriptionChanged,
+    } = this._oneSignalService.init();
+
+    const subscriptions = this._subscriptions;
+    subscriptions.push(notificationOpened.subscribe((value) => this.notificationOpened.emit(value)));
```
Comment (paulsouche) [inline @santech/angular-ionic/src/directives/one-signal.directive.ts]: subscriptions.push(sub1, sub2, sub3)

File: @santech/angular-ionic/src/directives/one-signal.directive.ts [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+    const pt = await this._platform.ready();
+    if (pt !== cordovaPlatform) {
+      return;
+    }
+
+    const {
+      notificationOpened,
+      notificationReceived,
+      subscriptionChanged,
+    } = this._oneSignalService.init();
```
Comment (paulsouche) [inline @santech/angular-ionic/src/directives/one-signal.directive.ts]: try {} => catch => onError.emit()

File: @santech/angular-ionic/src/directives/one-signal-authenticated.directive.ts [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+    this._platform = platform;
+    this._oneSignalService = oneSignalService;
+  }
+
+  public async ngOnInit() {
+    if (!await this._isCordova()) {
+      return;
+    }
+
+    return this._oneSignalService.enableNotifications()
```
Comment (paulsouche) [inline @santech/angular-ionic/src/directives/one-signal-authenticated.directive.ts]: async +catch onError emit

File: @santech/angular-ionic/src/directives/one-signal-authenticated.directive.ts [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+      .then((ids) => this.onNotificationsEnabled.emit(ids))
+      .catch(noop);
+  }
+
+  public async ngOnDestroy() {
+    if (!await this._isCordova()) {
+      return;
+    }
+
+    this._oneSignalService.disableNotifications();
```
Comment (paulsouche) [inline @santech/angular-ionic/src/directives/one-signal-authenticated.directive.ts]: async +catch onDestroyError emit

**Pull request merged**